### PR TITLE
Add option to disable half precision BLAS functions

### PR DESCRIPTION
**Pull request by sbalint98**
# Description
This PR adds functionality to disable the cuBLAS and MKLcpu BLAS functions, which take half-precision arguments. related to #99

# Checklist

- [ ] Do all unit tests pass locally? Attach a log.
- [x] Have you formatted the code using clang-format?

### CODE REVIEW

File: src/blas/backends/mklcpu/mklcpu_level3.cxx [hunk 1]
```diff
@@ -172,6 +177,9 @@ void gemm(cl::sycl::queue &queue, transpose transa, transpose transb, int64_t m,
             ::free(f32_b);
         });
     });
+#else
+    throw oneapi::mkl::unimplemented("blas", "cl::sycl::half", "when using hipSYCL");
```
Comment (mkrainiuk) [inline src/blas/backends/mklcpu/mklcpu_level3.cxx]: Please align this error message with more general one "half is disabled", I don't think it should be hipSYCL specific.

File: src/config.hpp.in [hunk 2]
```diff
@@ -25,6 +25,7 @@
 #cmakedefine ENABLE_MKLCPU_BACKEND
 #cmakedefine ENABLE_MKLGPU_BACKEND
 #cmakedefine ENABLE_NETLIB_BACKEND
+#cmakedefine ENABLE_HALF_ROUTINES
```
Comment (mkrainiuk) [inline src/config.hpp.in]: I think new flag needs to be applied to all backends, but I see changes only in cublas and mklcpu, is there a reason why it's not propagated to other backends?

**Review by mkrainiuk — COMMENTED** [review]
Thank you for PR! I have two comments for these changes.

File: src/config.hpp.in (hunk 2 above)
Comment (sbalint98) [reply src/config.hpp.in]: This seems to be an oversight on my part, I only tested with these two backends.

Comment (mmeterel) [reply src/config.hpp.in]: This probable means more changes needs to be done on other backends as well. 

However, I have my reservations about disabling half precision since it is part of SYCL Specification 1.2.1 section 5.1. Could you please remind me your argument for not supporting half precision in hipSYCL?

Comment (vrpascuzzi) [reply src/config.hpp.in]: > I have my reservations about disabling half precision since it is part of SYCL Specification 1.2.1 section 5.1.

Maybe more importantly, it is part of the SYCL 2020 spec, C.9.1.

Comment (vrpascuzzi) [reply src/config.hpp.in]: If it's necessary to disable `half` for hipSYCL, can disabling be done implicitly when using hipSYCL rather than adding an explicit functionality, instead of introducing more complexity into the build system?

Comment (illuhad) [reply src/config.hpp.in]: @mmeterel The SYCL specification lists half support as an optional kernel feature. It is not necessary for an implementation to support it inside kernels or other operations running on the device, but the type has to exist.

@sbalint98 Could we just use the SYCL 2020 API to query `half` support at runtime instead of hardcoding via the build system? https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:optional-kernel-features

This is something that should be done in general as a SYCL 2020 best practice. Half is an optional device feature, and therefore support in general should be queried before using it.

EDIT: Since the half type is only defined as part of the OpenCL backend specification [1], it could even be argued that the type may not have to exist for SYCL implementations that don't run on OpenCL. The spec is a bit unclear here.

[1] https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:opencl:extension-fp16

Comment (illuhad) [reply src/config.hpp.in]: @vrpascuzzi 

> Maybe more importantly, it is part of the SYCL 2020 spec, C.9.1.

No, appendix C is the OpenCL backend specification which only applies to SYCL implementations running on top of OpenCL and following the OpenCL backend specification. It neither applies to hipSYCL's backends (which are not OpenCL) nor to DPC++ Level Zero or DPC++ CUDA.

The more relevant section is here: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:optional-kernel-features

Comment (vrpascuzzi) [reply src/config.hpp.in]: Thanks for the info! Then the important part of that section is:
> In order to guarantee source code portability of SYCL applications that use optional kernel features, all SYCL implementations must be able to compile device code that uses these optional features regardless of whether the implementation supports the features on any of its devices.

So is it correct to disable entirely fp16 as is done here?

Comment (illuhad) [reply src/config.hpp.in]: I'm not sure. The spec is a bit unclear. The existence of the type is only mentioned in the OpenCL backend spec, but the intent indeed was to allow optional kernel features inside kernels if support is queried at runtime - this requires compile-time support. It's not very clear if optional types from a *backend specification* should be supported at compile-time for other backends.

hipSYCL has the `half` type (despite it not being functional), so it should be possible to compile the interface. I think we should investigate whether it is sufficient if we just test at runtime for the `fp16` device aspect, as I had proposed above.

Comment (vrpascuzzi) [reply src/config.hpp.in]: I think "_...all SYCL implementations..._" is very clear.

A run-time test seems fine from my point of view, so long as it compiles without modification to the source nor does it require additional CMake variables.

And by the way, thanks for hipSYCL!

Comment (illuhad) [reply src/config.hpp.in]: > I think "...all SYCL implementations..." is very clear.

It's not even defined in the core spec what the `fp16` optional feature means, only in the backend specification. At the same time, backend specifications are not binding for other backends. IMO there's a gap there. The definition of the features mentioned in the core spec (`fp16`, `fp64`, `atomic64`) should also be defined in the core spec if they are supposed to be well-defined everywhere.

> And by the way, thanks for hipSYCL!

'Tis our pleasure :)

Comment (vrpascuzzi) [reply src/config.hpp.in]: > It's not even defined in the core spec what the fp16 optional feature means...

Can it be that it's definition is inferred by the LLVM?
https://clang.llvm.org/docs/LanguageExtensions.html#half-precision-floating-point

AMDGPU is one of the currently supported targets. Out of curiosity (for my own understanding), what is the limitation in hipSYCL to support fp16? hipSYCL is also based on LLVM.

Comment (illuhad) [reply src/config.hpp.in]: > Can it be that it's definition is inferred by the LLVM?

My guess is that it's an oversight that happened when content was shuffled around when we generalized the SYCL backend model for SYCL 2020. In any case, LLVM conventions are also not binding for the SYCL spec :)

> AMDGPU is one of the currently supported targets. Out of curiosity (for my own understanding), what is the limitation in hipSYCL to support fp16? hipSYCL is also based on LLVM.

It's mainly an implementation prioritization issue. On GPU it's fairly straight forward as you point out, but on CPU we have to build some emulation or ship some C++ half library with hipSYCL.

Comment (vrpascuzzi) [reply src/config.hpp.in]: Perhaps it'd be useful to ask @bader or @Pennycook for clarification.

Thanks for the fruitful discussion, @illuhad! Good luck with this!

Comment (Pennycook) [reply src/config.hpp.in]: I agree that the specification is unclear, here.  The definition of `half` is something that I think needs more discussion.

There are already a few internal Khronos issues open that I think are relevant to this issue: https://gitlab.khronos.org/sycl/Specification/-/issues/322, https://gitlab.khronos.org/sycl/Specification/-/issues/455.  @illuhad, if you agree that those issues are relevant, would you mind adding some more details there?

**Comment by sbalint98** [comment]
Sorry for the long delay. Added runtime checking for the support for fp16 types.

File: include/oneapi/mkl/types.hpp [hunk 3]
```diff
@@ -102,6 +105,26 @@ enum class order : char {
     E = 1,
 };
 
+// Utility function to verify that a given set of types is supported by the
+// device compiler combintiation
+template<typename verify_type, typename T, typename... Ts>
+bool verify_support(cl::sycl::queue q, cl::sycl::aspect aspect){
```
Comment (mkrainiuk) [inline include/oneapi/mkl/types.hpp]: oneapi/mkl/types.hpp is a public header, as this function is more related to implementation detail, could you please move it to src/include?

Comment (sbalint98) [reply include/oneapi/mkl/types.hpp]: Sure, I added a new file `runtime_support_helper.hpp` to the `src/include` folder. I also had to modify the CMakeLists.txt files for the backends so that the `src/include` folder is also among the include directories. I added this change to all the BLAS backends. Should we add it to the other domains as well? Let me know what you think about this change and in case you don't want to have a new file for only this function where it would fit well.

Comment (mkrainiuk) [reply include/oneapi/mkl/types.hpp]: Thanks @sbalint98! So far only BLAS has half type API, so there is no need to add this header to other domains. New file looks good to me, I don't think it fits to any files that we have in this folder.

Could you please run clang-format for new file and `src/blas/backends/cublas/cublas_level3.cpp`?

Comment (sbalint98) [reply include/oneapi/mkl/types.hpp]: Sure. I keep forgetting that.

**Review by mkrainiuk — APPROVED** [review]

File: include/oneapi/mkl/types.hpp (hunk 3 above)
Comment (mkrainiuk) [reply include/oneapi/mkl/types.hpp]: Thanks!

File: src/blas/backends/mklcpu/mklcpu_level3.cxx [hunk 4]
```diff
@@ -131,7 +134,7 @@ void gemm(cl::sycl::queue &queue, transpose transa, transpose transb, int64_t m,
             copy_mat(accessor_c, MKLMAJOR, transpose::N, m, n, ldc, 0.0f, f32_c);
             ::cblas_sgemm(CBLASMAJOR, transa_, transb_, m, n, k, f32_alpha, f32_a, lda, f32_b, ldb,
                           f32_beta, f32_c, ldc);
-            // copy C back to half
+            // copy C back to cl::sycl::half
```
Comment (mmeterel) [inline src/blas/backends/mklcpu/mklcpu_level3.cxx]: Could you please revert this change?

File: src/include/runtime_support_helper.hpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+*******************************************************************************/
+
+#ifndef _ONEMKL_RUNTIME_SUPPORT_HELPER_HPP_
+#define _ONEMKL_RUNTIME_SUPPORT_HELPER_HPP_
+
+#include <CL/sycl.hpp>
+#include <type_traits>
+
+// Utility function to verify that a given set of types is supported by the
+// device compiler combintiation
```
Comment (mmeterel) [inline src/include/runtime_support_helper.hpp]: combination (typo)

**Comment by mmeterel** [comment]
@sbalint98 Sorry for the delay. Could you please resolve conflicts in src/blas/backends/mklcpu/mklcpu_level3.cxx

File: src/include/runtime_support_helper.hpp (hunk 5 above)
Comment (sbalint98) [reply src/include/runtime_support_helper.hpp]: thanks, fixed!

**Review by mmeterel — APPROVED** [review]
